src/utils/genetic_operators.cpp

```diff
@@ -0,0 +1,203 @@
+/* Copyright 2017-2020 PaGMO development team
+
+This file is part of the PaGMO library.
+
+The PaGMO library is free software; you can redistribute it and/or modify
+it under the terms of either:
+
+  * the GNU Lesser General Public License as published by the Free
+    Software Foundation; either version 3 of the License, or (at your
+    option) any later version.
+
+or
+
+  * the GNU General Public License as published by the Free Software
+    Foundation; either version 3 of the License, or (at your option) any
+    later version.
+
+or both in parallel, as here.
+
+The PaGMO library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+// MINGW-specific warnings.
+#if defined(__GNUC__) && defined(__MINGW32__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsuggest-attribute=pure"
+#endif
+
+#include <random>
+#include <utility>
+
+#include <pagmo/rng.hpp>
+#include <pagmo/types.hpp>
+#include <pagmo/utils/generic.hpp>
+
+namespace pagmo
+{
+
+namespace detail
+{
+
+// Implementation of the binary crossover.
+// Requires the distribution index eta_c in [1, 100], crossover probability p_cr  in[0,1] -> undefined algo behaviour
+// otherwise Requires dimensions of the parent and bounds to be equal -> out of bound reads. dim_i is the integer
+// dimension (integer alleles assumed at the end of the chromosome)
+
+std::pair<vector_double, vector_double> sbx_crossover(const vector_double &parent1, const vector_double &parent2,
+                                                      const std::pair<vector_double, vector_double> &bounds,
+                                                      vector_double::size_type dim_i, const double p_cr,
+                                                      const double eta_c, detail::random_engine_type &random_engine)
+{
+    // Decision vector dimensions
+    auto dim = parent1.size();
+    auto dim_c = dim - dim_i;
+    // Problem bounds
+    const vector_double &lb = bounds.first;
+    const vector_double &ub = bounds.second;
+    // declarations
+    double y1, y2, yl, yu, rand01, beta, alpha, betaq, c1, c2;
+    vector_double::size_type site1, site2;
+    // Initialize the child decision vectors
+    vector_double child1 = parent1;
+    vector_double child2 = parent2;
+    // Random distributions
+    std::uniform_real_distribution<> drng(0., 1.); // to generate a number in [0, 1)
+
+    // This implements a Simulated Binary Crossover SBX and applies it to the non integer part of the decision
+    // vector
+    if (drng(random_engine) < p_cr) {
```

Comment:
The code would look better if you do an early exit here as:
```c++
if (drng(random_engine) >= p_cr)
   return std::make_pair(std::move(child1), std::move(child2));
```
Excessive indentation makes it harder to read.

Reply:
I personally do not like code with multiple exit points makes it more difficult to analys. Buts thats taste.

Reply:
Fair enough! :)